### DISCUSSION TITLE
Where to place test fixtures for `@nrwl/js:lib`?

### DISCUSSION THREAD

**Discussion by mattfysh** [nrwl/nx#10971] [General]
If my fixture files don't end with ".spec.ts", they will be picked up during build and complain about `jest`

If they do end with ".spec.ts" and have no tests inside, the fixtures file will throw the error "Your test suite must contain at least one test."

I'm thinking of manually editing the tsconfig.lib.json and tsconfig.spec.json files to change this behaviour, but I'm not fully across how migrations work. Will manually editing files that were generated by the plugin cause any future issues with migrations?

**Comment by rfprod** [comment]
> Will manually editing files that were generated by the plugin cause any future issues with migrations?

@mattfysh I think there will be no issues if you name your files with fixtures like `*.fixture.ts` or `*.mock.ts` or `*.patch.ts` or whatever unique suffix you chose and exclude them in `tsconfig.lib.json`
```json
...
  "exclude": ["**/test-setup.ts", "**/*.spec.ts", "**/*.fixture.ts", "**/*.mock.ts", ...],
...
```

You can include you fixtures in the `test-setup.ts` file in your library root

```typescript
import { someLibraryMock } from './lib/testing/some-library.mock';
```

and modify the library jest config

```typescript
...
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
...
```

It may be worth putting your fixtures or mocks into a separate unit testing library if you will use them to test several libraries, e.g.

`test-setup.ts`
```typescript
import { someLibraryMock } from '@app/backend-unit-testing';
```

If you do the latter, the `tsconfig.lib.json` will be cleaner

```json
...
  "exclude": ["**/test-setup.ts", "**/*.spec.ts", ...],
...
```
